src/InstanceRepository.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+/**
+ * This file is part of the contentful.php package.
+ *
+ * @copyright 2015-2018 Contentful GmbH
+ * @license   MIT
+ */
+
+namespace Contentful\Delivery;
+
+use Contentful\Core\Resource\ResourceInterface;
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * InstanceRepository class.
+ *
+ * This class acts as a registry for current objects managed by the Client.
+ * It also abstracts access to objects stored in cache.
+ */
+class InstanceRepository
+{
+    /**
+     * @var string[]
+     */
+    private static $warmupTypes = [
+        'ContentType',
+        'Environment',
+        'Locale',
+        'Space',
+    ];
+
+    /**
+     * @var array[ResourceInterface[]]
+     */
+    private $resources = [
+        'Asset' => [],
+        'ContentType' => [],
+        'Entry' => [],
+        'Locale' => [],
+        'Space' => [],
+    ];
+
+    /**
+     * @var Client
+     */
+    private $client;
+
+    /**
+     * @var string
+     */
+    private $api;
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cacheItemPool;
+
+    /**
+     * @var bool
+     */
+    private $autoWarmup;
+
+    /**
+     * When retrieving from cache, there's the risk of an endless loop,
+     * so we keep track of the resources we're currently warming up, and
+     * skip the process if we track that we're already in the process of doing so.
+     *
+     * @var bool[]
+     */
+    private $warmupStack = [];
+
+    /**
+     * @param Client                 $client
+     * @param CacheItemPoolInterface $cacheItemPool
+     * @param bool                   $autoWarmup
+     */
+    public function __construct(Client $client, CacheItemPoolInterface $cacheItemPool, $autoWarmup)
+    {
+        $this->client = $client;
+        $this->api = $client->getApi();
+        $this->cacheItemPool = $cacheItemPool;
+        $this->autoWarmup = $autoWarmup;
+    }
+
+    /**
+     * Warm up the local resource repository with instances fetched from cache.
+     *
+     * @param string $type
+     * @param string $resourceId
+     */
+    private function warmUp($type, $resourceId)
+    {
+        $key = self::generateCacheKey($this->api, $type, $resourceId);
+        if (isset($this->warmupStack[$key]) || isset($this->resources[$type][$resourceId]) || !\in_array($type, self::$warmupTypes, true)) {
+            return;
+        }
+
+        $item = $this->cacheItemPool->getItem($key);
+        if ($item->isHit()) {
+            $this->warmupStack[$key] = true;
+            $this->resources[$type][$resourceId] = $this->client->parseJson($item->get());
+            unset($this->warmupStack[$key]);
+        }
+    }
+
+    /**
+     * @param string $type
+     * @param string $resourceId
+     *
+     * @return bool
+     */
+    public function has($type, $resourceId)
+    {
+        $this->warmUp($type, $resourceId);
```

Comment:
What happens if this is run on a collection of entries? Will it try to request each individual entry?